hyperdrive.go

```diff
@@ -113,3 +166,112 @@ func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Bl
 
 	r.Init()
 }
+
+// For a SignedBlock to be valid:
+// 1. must not be nil
+// 2. have valid blockTime, Round, and Height
+// 3. have a valid signature
+// 4. signatory belongs to the same shard
+// 5. parent header is the block at head of the shard's blockchain
+func (hyperdrive *hyperdrive) validateBlock(shardHash sig.Hash, signedBlock block.SignedBlock) bool {
+	if signedBlock.Block.Equal(block.Block{}) {
+		return false
+	}
+	if signedBlock.Time.After(time.Now()) {
+		return false
+	}
+	if signedBlock.Round < 0 || signedBlock.Height < 0 {
+		return false
+	}
+
+	// Verify the signatory of the signedBlock
+	signatory, err := hyperdrive.signer.Verify(signedBlock.Block.Header, signedBlock.Signature)
+	if err != nil {
+		return false
+	}
+	if !signatory.Equal(signedBlock.Signatory) {
+		return false
+	}
+	if !hyperdrive.isSignatoryInShard(shardHash, signatory) {
+		return false
+	}
+
+	// Is block.ParentHeader (at H) == block.Header (at H-1)?
+	if _, ok := hyperdrive.shardBlockchains[shardHash]; !ok {
+		return false
+	}
+	blockchain := hyperdrive.shardBlockchains[shardHash]
+	parent, ok := blockchain.Head()
```

Comment:
I have been thinking about this, and the potential out-of-order arrival of Blocks means this condition cannot be accurately verified here. We can leave it here for now, but please define a FIXME that should move this conditional check to the moment that the Block is actually going to be sent to the StateMachine (which means the verification of this condition must be in the Replica and perhaps, therefore, all verification should be done in the Replica).

Reply:
Agreed, will move verification to Replica.

Reply:
@divyakoshy as per our discussion let's implement this immediately for this PR.

Reply:
👍 Implemented.